torch/csrc/jit/passes/shape_analysis.h

```diff
@@ -9,6 +9,28 @@ namespace jit {
 
 struct Graph;
 
+struct propagation_error : std::exception {};
+
+class PropertyPropBase{
+  // Used for both Shape Propagation and Dtype/Device Propagation
+  public:
+  explicit PropertyPropBase(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {
+  }
+  virtual ~PropertyPropBase() = default;
+
+  void propagateBlock(Block* block, bool insert_expands = true);
```

Comment:
this isn't great, but I think I like the alternatives even less like 
a) moving `insert_expands` to be the part of analysis' state
b) duplicating/specializing `processLoop` for Shape Analysis vs Dtype and Device prop
c) splitting `propagateBlock(..., insert_expands)` into `propagateBlock` and `postPropagateBlock` 
or `propagateBlockLoop` and `propagateBlock`

@eellison wdyt ?